src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TemplateBuilder.cs

```diff
@@ -29,44 +31,48 @@ internal class TemplateBuilder
             _viewEngine = viewEngine;
             _viewContext = viewContext;
             _viewData = viewData;
-            _metadata = metadata;
+            _modelExplorer = modelExplorer;
             _htmlFieldName = htmlFieldName;
             _templateName = templateName;
             _readOnly = readOnly;
             _additionalViewData = additionalViewData;
+
+            _model = modelExplorer.Model;
+            _metadata = modelExplorer.Metadata;
         }
 
         public string Build()
         {
-            if (_metadata.ConvertEmptyStringToNull && string.Empty.Equals(_metadata.Model))
+            if (_metadata.ConvertEmptyStringToNull && string.Empty.Equals(_model))
             {
-                _metadata.Model = null;
+                _model = null;
             }
 
-            var formattedModelValue = _metadata.Model;
-            if (_metadata.Model == null && _readOnly)
+            var formattedModelValue = _model;
+            if (_model == null && _readOnly)
             {
                 formattedModelValue = _metadata.NullDisplayText;
             }
 
             var formatString = _readOnly ? _metadata.DisplayFormatString : _metadata.EditFormatString;
 
-            if (_metadata.Model != null && !string.IsNullOrEmpty(formatString))
+            if (_model != null && !string.IsNullOrEmpty(formatString))
             {
-                formattedModelValue = string.Format(CultureInfo.CurrentCulture, formatString, _metadata.Model);
+                formattedModelValue = string.Format(CultureInfo.CurrentCulture, formatString, _model);
             }
 
             // Normally this shouldn't happen, unless someone writes their own custom Object templates which
             // don't check to make sure that the object hasn't already been displayed
-            if (_viewData.TemplateInfo.Visited(_metadata))
+            if (_viewData.TemplateInfo.Visited(_modelExplorer))
             {
                 return string.Empty;
             }
 
-            var viewData = new ViewDataDictionary(_viewData, model: null)
+            // We need to copy the ModelExplorer to copy the model metadata. Otherwise we might
+            // lose track of the model type/property.
+            var viewData = new ViewDataDictionary(_viewData)
             {
-                Model = _metadata.Model,
-                ModelMetadata = _metadata
+                ModelExplorer = _modelExplorer,
```

Comment:
the default `ConvertEmptyStringToNull` is `true`.  would need to combine that default, a `string.Empty` model, and a template that uses `ViewData.Model` rather than `ViewData.TemplateInfo.FormattedModelValue`.

since this code is used for more than just invocation of our `Action<IHtmlHelper`>s, the easiest way to see the difference would be to write a DisplayTemplate/String.cshtml or EditorTemplate/String.cshtml file that special-cases `null`.